terminal/src/com/jediterm/terminal/ui/TerminalPanel.java

```diff
@@ -1079,11 +1120,11 @@ void drawCursor(String c, Graphics2D gfx, TextStyle style) {
       }
 
       CharBuffer buf = new CharBuffer(c);
-      int xCoord = x * myCharSize.width + getInsetX();
+      int xCoord = computexCoord(x, y) + getInsetX();
```

Comment:
Unfortunately, it will cost some performance, because now painting (height,width) panel would require approximately `(height * width * width) / 2` calls of `gfx.getFontMetrics(font).charWidth(<char>)`.

Reply:
> Yes, this is a must for non-monospaced fonts;

There is another point that can be cached after rendering. If Terminal does not receive keyboard messages or mouse messages or the size of the window changes, you can directly return the results of the last rendering. This can solve the performance problem.

For me, the rendering of non-monospaced fonts is necessary because it is impossible for our program to carry font files of dozens of megabytes.

Reply:
You can build jediterm.jar from your fork and use it as a workaround. When proper support for non-monospaced is added, you can remove your fork. Will it work for you?

Reply:
> You can build jediterm.jar from your fork and use it as a workaround. When proper support for non-monospaced is added, you can remove your fork. Will it work for you?

https://github.com/JetBrains/jediterm/pull/207#discussion_r663031434

Reply:
Sorry, not sure I understand: if your software would use jediterm.jar built from your fork, it would solve non-monospaced fonts rendering issues for your users. Right?

Reply:
Regarding this change, fixing non-monospaced fonts rendering issues as it was suggested in this PR would lead to degraded performance for all fonts. I agree that non-monospaced fonts rendering issues should be fixed, but probably the fix should be more elaborate to not degrade performance for monospaced fonts. No particular plans, but I'll think about it.

Reply:
> Regarding this change, fixing non-monospaced fonts rendering issues as it was suggested in this PR would lead to degraded performance for all fonts. I agree that non-monospaced fonts rendering issues should be fixed, but probably the fix should be more elaborate to not degrade performance for monospaced fonts. No particular plans, but I'll think about it.

Please do not close this PR. I should have time to fix performance issues in the next few months. Add double buffering and rendering cache to it.

Reply:
Sure, let's keep it open. Not sure double buffering would help here. It seems to me that it's more about improving algorithm of calculating x-position of a cell.

Reply:
Optimal calculation of X coordinate algorithm
Now it only accounts for 3.6% of the CPU in the entire drawing cycle

![ok](https://user-images.githubusercontent.com/43266206/124320627-cb81b580-dbae-11eb-8b7f-7aecbc15aeed.jpg)

Reply:
The algorithm has been optimized and now only takes up very little CPU